fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneHighlighting.java

```diff
@@ -50,127 +50,214 @@
+            if (!matched) {
+                matched = handlePrefixMatch(prefixTokens);
+            }
+            if (!matched) {
+                handleNonMatchToken();
+            }
+        }
+
+        public String search(final Set<String> matchedTokens, final Set<String> prefixTokens, final boolean allMatchingRequired) throws IOException {
+            while (it.next()) {
+                do {
+                    standardIt.next();
+                    if (standardIt.startOffset() < it.startOffset()) {
+                        // We're handling a stop word (we don't need to check if that word matches)
+                        handleNonMatchToken();
+                    } else {
+                        // We're handling a real word (We need to check if that word matches)
+                        handleToken(matchedTokens, prefixTokens);
+                    }
+                } while (standardIt.startOffset() < it.startOffset());
+            }
 
             if (allMatchingRequired && (matchedPrefixes.size() < prefixTokens.size() || (matchedInText.size() < matchedTokens.size()))) {
                 // Query text not actually found in document text. Return null
                 return null;
             }
 
             // Text was found. Return text
-            int endOffset = offsetAtt.endOffset();
-            if (upto < endOffset && !luceneQueryHighlightParameters.isCutSnippets()) {
+            if (!cutSnippets) {
+                addBeforeTokens();
                 addNonMatch(sb, text.substring(upto));
-            } else if (luceneQueryHighlightParameters.isCutSnippets()) {
-                while (!ends.isEmpty()) {
-                    addNonMatch(sb, ends.poll());
-                }
-                addNonMatch(sb, highlightedTextConnector);
+            }
+
+            if (snippetRunningBudget >= 0 && cutSnippets) {
+                addNonMatch(sb, " ...");
             }
             return sb.toString();
+        }
+    }
 
+    @Nullable
+    static String searchAllAndHighlight(@Nonnull String fieldName, @Nonnull Analyzer queryAnalyzer, @Nonnull String text,
+                                        @Nonnull Set<String> matchedTokens, @Nonnull Set<String> prefixTokens,
+                                        boolean allMatchingRequired,
+                                        @Nonnull LuceneScanQueryParameters.LuceneQueryHighlightParameters luceneQueryHighlightParameters,
+                                        @Nullable List<Pair<Integer, Integer>> highlightedPositions) {
+        try (TokenStream ts = queryAnalyzer.tokenStream(fieldName, new StringReader(text)) ;
+             StandardAnalyzer standardAnalyzer = new StandardAnalyzer() ;
+             TokenStream standardTs = standardAnalyzer.tokenStream(fieldName, new StringReader(text)) ;
+             TokenIterator it = new TokenIterator(ts);
+             TokenIterator standardIt = new TokenIterator(standardTs)
+        ) {
```

Comment:
Does the `StandardAnalyzer` work for e.g. RTL languages? I think just a comment would be helpful.